### PR TITLE
Ea/story25

### PR DESCRIPTION
**Pull request by eamouzou**
- Pets Application epic finished
- Completed user stories 16 - 25
- Add tests, routes, controller actions, views, and migrations for application functionality

### CODE REVIEW

File: app/controllers/pets_controller.rb [hunk 1]
```diff
@@ -37,4 +37,19 @@ def destroy
...
+      pet.update(adoption_status: "pending")
+      pet_app = PetApplication.where(pet_id: params[:petid], application_id: params[:appid])[0]
+      pet_app.update(adopted: true)
+      redirect_to "/pets/#{pet.id}"
+    elsif pet.adoption_status == "pending"
+      pet.update(adoption_status: "adoptable")
+      pet_app = PetApplication.where(pet_id: params[:petid], application_id: params[:appid])[0]
+      pet_app[:adopted] = false
+      redirect_to "/applications/#{params[:appid]}"
+    end
```
Comment (eamouzou) [inline app/controllers/pets_controller.rb]: This process changes depending on whether the pet is already adopted or not. It toggles between rejecting the application or approving it; and it changes the adoption status of the pet accordingly.

File: app/models/pet.rb [hunk 2]
```diff
@@ -2,11 +2,21 @@ class Pet < ApplicationRecord
...
   def self.favpets(favorites)
     favorites.pets_favorite.keys.map do |petid|
       Pet.find(petid)
     end
   end
 
+  def default_values
+    self.adoption_status = "adoptable" if self.adoption_status.nil?
+    self.description = "Looking for a home" if self.description.nil?
+  end
```
Comment (eamouzou) [inline app/models/pet.rb]: sets a default of both of these attributes since we don't have validations

File: app/views/applications/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    <p>Describe why you would make a good home: <%= @application.description %></p>
+    <p>Pets applied for:</p>
+    <% @application.pets.each do |pet| %>
+      <section id= <%="pet#{pet.id}"  %>>
+      <p><%= link_to pet.name, "/pets/#{pet.id}" %></p>
+      <% if pet.adoption_status == "adoptable"  %>
+        <p><%= link_to "Approve #{pet.name}'s application", "/pets/#{pet.id}/applications/#{@application.id}/approve", method: :patch %></p>
+      <% elsif pet.adoption_status == "pending" %>
+        <p><%= link_to "Unapprove #{pet.name}'s application", "/pets/#{pet.id}/applications/#{@application.id}/reject", method: :patch %></p>
+      <% end %>
```
Comment (eamouzou) [inline app/views/applications/show.html.erb]: Depending on adoption status, it provides a link to approve or reject the application.

File: config/routes.rb [hunk 4]
```diff
@@ -32,4 +32,8 @@
 
   get '/applications/new', to: 'applications#new'
   post '/applications', to: 'applications#create'
+  get '/applications/:application_id', to: 'applications#show'
+  get '/pets/:id/applications', to: 'applications#index'
+  patch '/pets/:petid/applications/:appid/approve', to: 'pets#change_adoption'
+  patch '/pets/:petid/applications/:appid/reject', to: 'pets#change_adoption'
```
Comment (eamouzou) [inline config/routes.rb]: not sure if these belong here or in the pet block (feel free to move and adjust as you see fit)

File: db/migrate/20200206151814_create_pet_applications.rb [hunk 5]
```diff
@@ -3,6 +3,7 @@ def change
     create_table :pet_applications do |t|
       t.references :pet, foreign_key: true
       t.references :application, foreign_key: true
+      t.boolean :adopted, default: false
```
Comment (eamouzou) [inline db/migrate/20200206151814_create_pet_applications.rb]: added this to the pet table to facilitate the changing of adoption status

File: app/models/pet.rb (hunk 2 above)
Comment (Will-Meighan) [reply app/models/pet.rb]: Interesting. I haven't seen before_save before. I will have to look into it and might have you give me a little explanation when we meet up again.

**Comment by Will-Meighan** [comment]
Thanks for the hard work. Going to jump into the next several user stories now.

**Pull request merged**